source/common/upstream/eds.cc

```diff
@@ -1,53 +1,54 @@
-#include "common/upstream/sds.h"
+#include "common/upstream/eds.h"
 
-#include <cstdint>
-#include <string>
-#include <vector>
-
-#include "common/http/headers.h"
-#include "common/json/config_schemas.h"
 #include "common/network/address_impl.h"
 #include "common/network/utility.h"
+#include "common/upstream/sds_subscription.h"
 
 namespace Envoy {
 namespace Upstream {
 
-SdsClusterImpl::SdsClusterImpl(const Json::Object& config, Runtime::Loader& runtime,
+EdsClusterImpl::EdsClusterImpl(const Json::Object& config, Runtime::Loader& runtime,
                                Stats::Store& stats, Ssl::ContextManager& ssl_context_manager,
                                const SdsConfig& sds_config, const LocalInfo::LocalInfo& local_info,
                                ClusterManager& cm, Event::Dispatcher& dispatcher,
                                Runtime::RandomGenerator& random)
-    : BaseDynamicClusterImpl(config, runtime, stats, ssl_context_manager),
-      RestApiFetcher(cm, sds_config.sds_cluster_name_, dispatcher, random,
-                     sds_config.refresh_delay_),
-      local_info_(local_info), service_name_(config.getString("service_name")) {}
-
-void SdsClusterImpl::parseResponse(const Http::Message& response) {
-  Json::ObjectSharedPtr json = Json::Factory::loadFromString(response.bodyAsString());
-  json->validateSchema(Json::Schema::SDS_SCHEMA);
+    : BaseDynamicClusterImpl(config, runtime, stats, ssl_context_manager), local_info_(local_info),
+      cluster_name_(config.getString("service_name")) {
+  // TODO(htuch): This is where the v2 Subscription implementations will be constructed based on
+  // config. Today, we just reuse the v1 SDS config and an adapter Subscription to the v1 REST JSON
+  // API.
+  subscription_.reset(new SdsSubscription(info_->stats(), sds_config, cm, dispatcher, random));
+}
+
+void EdsClusterImpl::initialize() { subscription_->start({cluster_name_}, *this); }
+
+bool EdsClusterImpl::onConfigUpdate(const ResourceVector& resources) {
   std::vector<HostSharedPtr> new_hosts;
-  for (const Json::ObjectSharedPtr& host : json->getObjectArray("hosts")) {
-    bool canary = false;
-    uint32_t weight = 1;
-    std::string zone = "";
-    if (host->hasObject("tags")) {
-      canary = host->getObject("tags")->getBoolean("canary", canary);
-      weight = host->getObject("tags")->getInteger("load_balancing_weight", weight);
-      zone = host->getObject("tags")->getString("az", zone);
+  for (const auto& cluster_load_assignment : resources) {
```

Comment:
It really depends on the management server. I didn't want to make any code assumptions that could result in something bad (e.g. dereferencing an empty vector at element 0) when the management server is in control of inputs.

I'll add the assert and restructure a little to reflect the expected normal behavior.